### PR TITLE
fix(gitlab): smart truncate comments

### LINKED ISSUE #30551: platform(gitlab): comment length not truncated
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

Essentially https://github.com/renovatebot/renovate/issues/7775 has come back after some refactors. I think that we need a call to smartTruncate around here

https://github.com/renovatebot/renovate/blob/ad7b3ceda800812356b17d7183ced251cd279fc3/lib/modules/platform/gitlab/index.ts#L1405

So you could add a line like `body = smartTruncate(body, maxBodyLength());`

`smartTruncate` is already called in `ensureIssue` but not in `ensureComment`. I think this was just an oversight but it does cause renovate to sometimes crash for us in mysterious ways to our internal users until we inspect the underlying renovate logs. 

### Logs (if relevant)

<details><summary>Logs</summary>

```
"message": "Response code 400 (Bad Request)",
             "stack": "HTTPError: Response code 400 (Bad Request)\n    at Request.<anonymous> (/usr/lib/node_modules/renovate/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
             "options": {
               "headers": {
                 "user-agent": "RenovateBot/37.440.7 (https://github.com/renovatebot/renovate)",
                 "accept": "application/json",
                 "authorization": "***********",
                 "content-type": "application/json",
                 "content-length": "1001123",
                 "accept-encoding": "gzip, deflate, br"
               },
               "url": "REDACTED/merge_requests/3149/notes",
               "hostType": "gitlab",
               "username": "",
               "password": "",
               "method": "POST",
               "http2": false
             },
             "response": {
               "statusCode": 400,
               "statusMessage": "Bad Request",
               "body": {
                 "message": "400 Bad request - Note {:note=>[\"is too long (maximum is 1000000 characters)\"]}"
               },
               "headers": {...

```

</details>

_Originally posted by @henry-fn in https://github.com/renovatebot/renovate/discussions/30508_

### PR DESCRIPTION
**Pull request by amishk783**
## Context
- closes #30551

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by RahulGautamSingh — CHANGES_REQUESTED** [review]
Trunacate the body content before we compute commentNeedsUpdating. Else, we will compare full body content with comment body we fetch from api. And, the fetched body will be truncated whereas full body (from fn param) won't. 

This will cause the comment to be updated each time.

**Review by viceice — CHANGES_REQUESTED** [review]
See conversation

**Comment by amishk783** [comment]
Hi Rahul, Thanks for the guidance. I’ve moved smartTruncate(body, maxBodyLength()) before the commentNeedsUpdating check in the if-else case and will push the changes shortly.

**Comment by henry-fn** [comment]
Thanks @amishk783 for the fix, anything blocking this being merged in?

**Comment by viceice** [comment]
> Thanks @amishk783 for the fix, anything blocking this being merged in?

required changes are not pushed. also CI falls, so merge isn't an option.

**Comment by amishk783** [comment]
> > Thanks @amishk783 for the fix, anything blocking this being merged in?
> 
> required changes are not pushed. also CI falls, so merge isn't an option.

Hey, sorry for the delay. it completely slipped my mind. I've pushed the changes and addressed the feedback. Now truncating the body before checking commentNeedsUpdating

**Comment by henry-fn** [comment]
> > > Thanks @amishk783 for the fix, anything blocking this being merged in?
> > 
> > 
> > required changes are not pushed. also CI falls, so merge isn't an option.
> 
> Hey, sorry for the delay. it completely slipped my mind. I've pushed the changes and addressed the feedback. Now truncating the body before checking commentNeedsUpdating

Just in case you missed it, I think there is still one linting test failing https://github.com/renovatebot/renovate/actions/runs/10960183807/job/30434310107?pr=30580

**Comment by henry-fn** [comment]
@amishk783 poke on formatting here?

**Review by rarkins — APPROVED** [review]

**Comment by henry-fn** [comment]
Anything blocking merging here?

**Comment by rarkins** [comment]
@viceice please review

**Comment by henry-fn** [comment]
Checking in again here

File: lib/modules/platform/gitlab/index.ts [hunk 1]
```diff
@@ -1400,13 +1401,15 @@ export async function ensureComment({
   } else {
     logger.debug(`Ensuring content-only comment in #${number}`);
     body = `${sanitizedContent}`;
+    body = smartTruncate(body, maxBodyLength());
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1404]:
```suggestion
    body = smartTruncate(`${sanitizedContent}`, maxBodyLength());
```

File: lib/modules/platform/gitlab/index.ts [hunk 2]
```diff
@@ -1391,6 +1391,7 @@ export async function ensureComment({
     body = body
       .replace(regEx(/Pull Request/g), 'Merge Request')
       .replace(regEx(/PR/g), 'MR');
+    body = smartTruncate(body, maxBodyLength());
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1394]:
```suggestion
     body = smartTruncate(body
       .replace(regEx(/Pull Request/g), 'Merge Request')
       .replace(regEx(/PR/g), maxBodyLength());
```
needs prettifying

**Comment by jwilliams-fn** [comment]
@amishk783 poking here for _hopefully_ the final time 😅

**Comment by RahulGautamSingh** [comment]
Replacement PR: https://github.com/renovatebot/renovate/pull/33348

**Pull request closed without merging**